fileutil/tarball_compressor.go

```diff
@@ -32,44 +37,130 @@ func (c tarballCompressor) CompressSpecificFilesInDir(dir string, files []string
 
 	defer tarball.Close()
 
-	tarballPath := tarball.Name()
+	zw := gzip.NewWriter(tarball)
+	tw := tar.NewWriter(zw)
+
+	for _, file := range files {
+		err = c.fs.Walk(filepath.Join(dir, file), func(f string, fi os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if filepath.Base(f) == ".DS_Store" {
+				return nil
+			}
+
+			header, err := tar.FileInfoHeader(fi, f)
+			if err != nil {
+				return bosherr.WrapError(err, "Reading tar header")
+			}
+
+			relPath, err := filepath.Rel(dir, filepath.ToSlash(f))
+			if err != nil {
+				return bosherr.WrapError(err, "Resovling relative tar path")
+			}
+			header.Name = relPath
+
+			if err := tw.WriteHeader(header); err != nil {
+				return bosherr.WrapError(err, "Writing tar header")
+			}
+
+			if fi.Mode().IsRegular() {
+				data, err := os.Open(f)
+				if err != nil {
+					return bosherr.WrapError(err, "Reading tar source file")
+				}
+				if _, err := io.Copy(tw, data); err != nil {
+					return bosherr.WrapError(err, "Copying data into tar")
+				}
+			}
+			return nil
+		})
+	}
 
-	args := []string{"-czf", tarballPath, "-C", dir}
-	if runtime.GOOS == "darwin" {
-		args = append([]string{"--no-mac-metadata"}, args...)
+	if err != nil {
+		return "", bosherr.WrapError(err, "Creating tgz")
 	}
 
-	for _, file := range files { //nolint:gosimple
-		args = append(args, file)
+        if err = tw.Close(); err != nil {
+		return "", bosherr.WrapError(err, "Closing tar writer")
 	}
 
-	_, _, _, err = c.cmdRunner.RunCommand("tar", args...)
-	if err != nil {
-		return "", bosherr.WrapError(err, "Shelling out to tar")
+	if err = zw.Close(); err != nil {
+		return "", bosherr.WrapError(err, "Closing gzip writer")
 	}
 
-	return tarballPath, nil
+	return tarball.Name(), nil
 }
 
 func (c tarballCompressor) DecompressFileToDir(tarballPath string, dir string, options CompressorOptions) error {
-	sameOwnerOption := "--no-same-owner"
-	if options.SameOwner {
-		sameOwnerOption = "--same-owner"
+	if _, err := c.fs.Stat(dir); os.IsNotExist(err) {
+		return bosherr.WrapError(err, "Determine target dir")
 	}
 
-	args := []string{sameOwnerOption, "-xzf", tarballPath, "-C", dir}
-	if options.StripComponents != 0 {
-		args = append(args, fmt.Sprintf("--strip-components=%d", options.StripComponents))
+	tarball, err := c.fs.OpenFile(tarballPath, os.O_RDONLY, 0)
+	if err != nil {
+		return bosherr.WrapError(err, "Opening tarball")
 	}
 
-	if options.PathInArchive != "" {
-		args = append(args, options.PathInArchive)
-	}
-	_, _, _, err := c.cmdRunner.RunCommand("tar", args...)
+	zr, err := gzip.NewReader(tarball)
```

Comment:
Should we add `defer tarball.Close()` and `defer zr.Close()`? The `CompressFilesInDir` function is closing the tarball stream e.g.. Or is the close handled differently here because I couldn't follow where the close is happening?